### ISSUE TITLE
[ENHANCEMENT]: Consider removing the use of `argument_type` and `result_type` from `hasher` in cuco `default_filter_policy`

### ISSUE THREAD

**Issue by mhaseeb123** [NVIDIA/cuCollections#653]
### Is your feature request related to a problem? Please describe.

The member type std::hash<T>::argument_type is deprecated in C++17 and removed in C++20, so we should consider syncing with STL and remove them from our hashers as well as their use in bloom filter policies.

Ref: https://en.cppreference.com/w/cpp/utility/hash

_Originally posted by @bdice in https://github.com/rapidsai/cudf/pull/17289#discussion_r1876965059_

### Describe the solution you'd like

Consider refactoring to not use hasher::argument_type and hasher::result_type across cuCollections.

### Describe alternatives you've considered

Keep using the deprecated hasher type aliases until C++20

**Comment by PointKernel** [comment]
I thought about mentioning this during the code review but didn’t want to come across as nitpicking. Instead of https://github.com/NVIDIA/cuCollections/blob/b8429d44314a02da9b5ebc17fa0a9284543aa825/include/cuco/detail/bloom_filter/arrow_filter_policy.cuh#L84-L85

we could make the hasher a template type:
```cpp
template<class Key, template <typename> class Hash>
class arrow_filter_policy {
  ...
};
```

To use it, users can simply pass the hasher type without repeating the `Key` type again, e.g.
```cpp
auto p = arrow_filter_policy<Key, cuco::xxhash_32>{...};
```

**Comment by mhaseeb123** [comment]
> I thought about mentioning this during the code review but didn’t want to come across as nitpicking.

No worries about that. I am always for pursuing excellence and attention to the detail is how we all learn. 😄 

I think this is a great idea. Let's do this. But we still would be using the member types (`argument_type` and `result_type`) from the hasher which Bradley mentioned

**Comment by PointKernel** [comment]
>  we still would be using the member types (argument_type and result_type) from the hasher

Half-half. The `argument_type` will simply be `key_type (using key_type = Key)`. However, retaining `result_type` (should be `hash_value_type` of the policy) is still desirable. The STL simplifies hashers by always using `std::size_t` as the return type, but this approach is suboptimal, especially in terms of performance and for GPUs. Such oversimplification does not fit our goals and IMO should be avoided.

**Comment by mhaseeb123** [comment]
> Such oversimplification does not fit our goals and IMO should be avoided.

Agreed. Let's make the suggested change to `cuco::arrow_filter_policy` and close this issue otherwise.

**Comment by mhaseeb123** [comment]
#654 removes these member types for the `arrow_filter_policy`.
